internal/typeinfo/member.go

```diff
@@ -0,0 +1,99 @@
+package typeinfo
+
+import (
+	"database/sql"
+	"fmt"
+	"reflect"
+)
+
+var scannerInterface = reflect.TypeOf((*sql.Scanner)(nil)).Elem()
+
+// Member describes a type that is a child
+// of some other encapsulating type.
+type Member interface {
+	// OuterType returns the outer type in which this member is present.
+	OuterType() reflect.Type
+
+	// MemberName returns this member's name.
+	MemberName() string
+
+	// ValueForOuter returns the value represented by this
+	// type member within the input outer value.
+	ValueForOuter(reflect.Value) (reflect.Value, error)
+
+	AddScanTarget(reflect.Value, []any, []ScanProxy) ([]any, []ScanProxy, error)
+}
+
+type mapKey struct {
+	name    string
+	mapType reflect.Type
+}
+
+func (mk *mapKey) OuterType() reflect.Type {
+	return mk.mapType
+}
+
+func (mk *mapKey) MemberName() string {
+	return mk.name
+}
+
+func (mk *mapKey) ValueForOuter(v reflect.Value) (reflect.Value, error) {
+	val := v.MapIndex(reflect.ValueOf(mk.name))
+	if val.Kind() == reflect.Invalid {
+		return val, fmt.Errorf("map %q does not contain key %q", mk.OuterType().Name(), mk.name)
+	}
+	return val, nil
+}
+
+func (mk *mapKey) AddScanTarget(outVal reflect.Value, ptrs []any, proxies []ScanProxy) ([]any, []ScanProxy, error) {
```

Comment:
It would be good to note the reason that we need the `ScanProxy` here, namely that Go does not expose pointers to map values (this would maybe be an internal comment rather than on the public function).